### PR TITLE
[RISCV][TTI] Model the cost of insert/extractelt when the vector split into multiple register group and idx exceed single group.

### PR DESCRIPTION
**Pull request by ElvisWang123**
This patch implements the cost when the size of the vector need to split into multiple groups and the index exceed single vector group.

For extract element, we need to store split vectors to stack and load the target element.
For insert element, we need to store split vectors to stack and store the target element and load vectors back.

After this patch, the cost of insert/extract element will close to the generated assembly.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 1]
```diff
@@ -1945,6 +1945,23 @@ InstructionCost RISCVTTIImpl::getVectorInstrCost(unsigned Opcode, Type *Val,
     // TODO: should we count these special vsetvlis?
     BaseCost = Opcode == Instruction::InsertElement ? 3 : 4;
   }
+
+  // When the vector need to split into multiple register groups and the index
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1949]: need -> needs

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 2]
```diff
@@ -1945,6 +1945,23 @@ InstructionCost RISCVTTIImpl::getVectorInstrCost(unsigned Opcode, Type *Val,
...
+  // exceed single vector resgister group, we need to extract the element via
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1950]: exceed -> exceeds

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1950]: resgister -> register

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 1 above)
Comment (ElvisWang123) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Fixed, thanks.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 3]
```diff
@@ -1922,6 +1922,33 @@ InstructionCost RISCVTTIImpl::getVectorInstrCost(unsigned Opcode, Type *Val,
...
+  if (LT.first > 1 &&
+      ((Index == -1U) || (Index >= LT.second.getVectorMinNumElements() &&
+                          LT.second.isScalableVector()))) {
+    Type *ScalarType = Val->getScalarType();
+    Align VecAlign = DL.getPrefTypeAlign(Val);
+    Align SclAlign = DL.getPrefTypeAlign(ScalarType);
+
+    // Store all split vectors into stack and load the target element.
+    if (Opcode == Instruction::ExtractElement)
+      return LT.first * getMemoryOpCost(Instruction::Store, Val, VecAlign, 0,
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1937]: You shouldn't need to multiply by LT.first here.  getMemoryOpCost does that internally.  Or, at least, it should.  If it's not doing so in the split case, we should fix that in getMemoryOpCost instead.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 4]
```diff
@@ -1922,6 +1922,33 @@ InstructionCost RISCVTTIImpl::getVectorInstrCost(unsigned Opcode, Type *Val,
...
+                                        CostKind) +
+             getMemoryOpCost(Instruction::Load, ScalarType, SclAlign, 0,
+                             CostKind);
+
+    // Store all split vectors into stack and store the target element and load
+    // vectors back.
+    return LT.first * (getMemoryOpCost(Instruction::Store, Val, VecAlign, 0,
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1944]: Same point here.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 5]
```diff
@@ -1922,6 +1922,33 @@ InstructionCost RISCVTTIImpl::getVectorInstrCost(unsigned Opcode, Type *Val,
...
+                          LT.second.isScalableVector()))) {
+    Type *ScalarType = Val->getScalarType();
+    Align VecAlign = DL.getPrefTypeAlign(Val);
+    Align SclAlign = DL.getPrefTypeAlign(ScalarType);
+
+    // Store all split vectors into stack and load the target element.
+    if (Opcode == Instruction::ExtractElement)
+      return LT.first * getMemoryOpCost(Instruction::Store, Val, VecAlign, 0,
+                                        CostKind) +
+             getMemoryOpCost(Instruction::Load, ScalarType, SclAlign, 0,
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1939]: You're missing the addressing cost in both cases here.  For the vector, that should be handled inside getMemoryOpCost, but you need to include the ADDI for the non-constant index case on the scalar load or store.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 6]
```diff
@@ -1945,6 +1972,7 @@ InstructionCost RISCVTTIImpl::getVectorInstrCost(unsigned Opcode, Type *Val,
     // TODO: should we count these special vsetvlis?
     BaseCost = Opcode == Instruction::InsertElement ? 3 : 4;
   }
+
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1975]: Remove stray whitespace.

File: llvm/test/Analysis/CostModel/RISCV/rvv-extractelement.ll [hunk 7]
```diff
@@ -926,10 +926,10 @@ define void @extractelement_int_lmul(i32 %x) {
 ; RV64ZVE64X-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %v32i32_15 = extractelement <32 x i32> undef, i32 15
 ; RV64ZVE64X-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %v32i32_31 = extractelement <32 x i32> undef, i32 31
 ; RV64ZVE64X-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %v64i32_63 = extractelement <64 x i32> undef, i32 63
-; RV64ZVE64X-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %v128i8 = extractelement <128 x i8> undef, i32 %x
-; RV64ZVE64X-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %v256i8 = extractelement <256 x i8> undef, i32 %x
-; RV64ZVE64X-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %v32i32 = extractelement <32 x i32> undef, i32 %x
-; RV64ZVE64X-NEXT:  Cost Model: Found an estimated cost of 2 for instruction: %v64i32 = extractelement <64 x i32> undef, i32 %x
+; RV64ZVE64X-NEXT:  Cost Model: Found an estimated cost of 33 for instruction: %v128i8 = extractelement <128 x i8> undef, i32 %x
+; RV64ZVE64X-NEXT:  Cost Model: Found an estimated cost of 129 for instruction: %v256i8 = extractelement <256 x i8> undef, i32 %x
```
Comment (preames) [inline llvm/test/Analysis/CostModel/RISCV/rvv-extractelement.ll:930]: This cost seems way too big.

256 x i8 on V is 16 registers worth of data.  Storing that should be ~16 in cost.  A single scalar load and addressing is ~2-3.  So, I'd expect something in the order of ~20, not ~129 here.

**Review by preames — CHANGES_REQUESTED** [review]

**Comment by preames** [comment]
> CodeGen for extract/insert element exceed single register group. https://godbolt.org/z/r1zKEoK1P

Thank you for including this.  It made spot checking the codegen much simpler.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 3 above)
Comment (ElvisWang123) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Removed, thanks!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 4 above)
Comment (ElvisWang123) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Removed, thanks!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 5 above)
Comment (ElvisWang123) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:1998]: Added, thanks!

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 6 above)
Comment (ElvisWang123) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Removed, thanks.

File: llvm/test/Analysis/CostModel/RISCV/rvv-extractelement.ll (hunk 7 above)
Comment (ElvisWang123) [reply llvm/test/Analysis/CostModel/RISCV/rvv-extractelement.ll]: I think  256 x i8 need 32 register under zve64x. 256 / 64 * 8 = 32
The cost of 128 x i8 is `18` and that is close to your expectation.

**Review by preames — APPROVED** [review]
LGTM

**Pull request merged**